File: deploy/new_html/services/entityFileService.ts
import { apiJson } from './httpClient';

export interface EntityFile {
  fileId: string;
  fileUrl: string;
  fileType: string;
  fileRole: string;
  isSelected: boolean;
  createdAt: string;
  metadata?: Record<string, unknown>;
  entityType?: string;
  entityId?: string;
}

function normalize(row: any): EntityFile {
  return {
    fileId: row.file_id ?? row.fileId ?? '',
    fileUrl: row.file_url ?? row.fileUrl ?? '',
    fileType: row.file_type ?? row.fileType ?? '',
    fileRole: row.file_role ?? row.fileRole ?? '',
    isSelected: !!(row.is_selected ?? row.isSelected),
    createdAt: row.created_at ?? row.createdAt ?? '',
    metadata: row.metadata,
    entityType: row.entity_type ?? row.entityType,
    entityId: row.entity_id ?? row.entityId,
  };
}

export async function fetchEntityFiles(
  entityType: string,
  entityId: string,
  fileRole?: string,
): Promise<{ items: EntityFile[]; total: number }> {
  const params = new URLSearchParams({ entity_type: entityType, entity_id: entityId });
  if (fileRole) params.set('file_role', fileRole);
  const data = await apiJson<{ items?: any[]; total?: number }>(
    `/api/entity-files?${params}`,
    { method: 'GET' },
    'fetchEntityFiles',
  );
  return {
    items: (data.items || []).map(normalize),
    total: data.total ?? 0,
  };
}

export async function fetchUserFiles(
  fileType?: string,
  limit: number = 100,
  offset: number = 0,
): Promise<{ items: EntityFile[]; total: number }> {
  const params = new URLSearchParams();
  if (fileType) params.set('file_type', fileType);
  params.set('limit', String(limit));
  params.set('offset', String(offset));
  const data = await apiJson<{ items?: any[]; total?: number }>(
    `/api/user-files?${params}`,
    { method: 'GET' },
    'fetchUserFiles',
  );
  return {
    items: (data.items || []).map(normalize),
    total: data.total ?? 0,
  };
}

export async function selectEntityFile(
  fileId: string,
  entityType: string,
  entityId: string,
  fileRole: string,
): Promise<EntityFile> {
  const data = await apiJson<{ file: any }>(`/api/entity-files/${fileId}/select`, {
    method: 'PUT',
    body: JSON.stringify({ entity_type: entityType, entity_id: entityId, file_role: fileRole }),
  }, 'selectEntityFile');
  return normalize(data.file);
}

export async function deleteEntityFile(fileId: string): Promise<void> {
  await apiJson<{ success: boolean }>(`/api/entity-files/${fileId}`, {
    method: 'DELETE',
  }, 'deleteEntityFile');
}

export async function hardDeleteEntityFile(fileId: string): Promise<{ freed_bytes: number }> {
  return apiJson<{ freed_bytes: number }>(`/api/entity-files/${fileId}/hard`, {
    method: 'DELETE',
  }, 'hardDeleteEntityFile');
}

export async function hardDeleteEntityFiles(fileIds: string[]): Promise<{ deleted: number; freed_bytes: number; errors: string[] }> {
  return apiJson<{ deleted: number; freed_bytes: number; errors: string[] }>('/api/entity-files/hard-delete-batch', {
    method: 'POST',
    body: JSON.stringify({ file_ids: fileIds }),
  }, 'hardDeleteEntityFiles');
}

export async function linkEntityFile(
  fileId: string,
  entityType: string,
  entityId: string,
  fileRole: string,
  isSelected: boolean = false,
): Promise<EntityFile> {
  const data = await apiJson<{ file: any }>('/api/entity-files/link', {
    method: 'POST',
    body: JSON.stringify({
      file_id: fileId, entity_type: entityType,
      entity_id: entityId, file_role: fileRole, is_selected: isSelected,
    }),
  }, 'linkEntityFile');
  return normalize(data.file);
}

export async function uploadEntityFile(
  file: File,
  entityType: string,
  entityId: string,
  fileRole: string,
  episodeId?: string,
): Promise<{ fileId: string; fileUrl: string }> {
  const formData = new FormData();
  formData.append('file', file);
  if (entityType) formData.append('entity_type', entityType);
  if (entityId) formData.append('entity_id', entityId);
  if (fileRole) formData.append('file_role', fileRole);
  if (episodeId) formData.append('episode_id', episodeId);

  const data = await apiJson<{ file_id: string; file_url: string }>('/api/entity-files/upload', {
    method: 'POST',
    body: formData,
  }, 'uploadEntityFile', { includeContentType: false });
  return { fileId: data.file_id, fileUrl: data.file_url };
}

File: deploy/services/entity_file_service.py
"""Entity file business logic for shared file attachments."""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Callable, Dict, Iterable, Optional


class EntityFileServiceError(RuntimeError):
    pass


class EntityFileNotFound(EntityFileServiceError):
    pass


class EntityFileBatchTooLarge(EntityFileServiceError):
    pass


class EntityFileDeleteRiskNotAcknowledged(EntityFileServiceError):
    pass


class EntityFilePhysicalDeleteFailed(EntityFileServiceError):
    pass


class EntityFileMigrationFailed(EntityFileServiceError):
    pass


def _normalize_file_rows(rows: Iterable[Any]) -> list[Dict[str, Any]]:
    items = []
    for row in rows:
        item = dict(row)
        if isinstance(item.get("metadata"), str):
            try:
                item["metadata"] = json.loads(item["metadata"])
            except Exception:
                item["metadata"] = {}
        items.append(item)
    return items


def _detect_file_type(content_type: Optional[str]) -> str:
    if content_type and content_type.startswith("image"):
        return "image"
    if content_type and content_type.startswith("audio"):
        return "audio"
    if content_type and content_type.startswith("video"):
        return "video"
    return "other"


async def list_user_files(
    *,
    user_id: str,
    file_type: Optional[str],
    limit: int,
    offset: int,
    file_dao: Any,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    capped_limit = min(limit, 500)
    rows = await file_dao.get_user_files(user_id, file_type, capped_limit, offset)
    total = await entity_file_dao.count_user_files(user_id, file_type)
    return {"success": True, "items": _normalize_file_rows(rows), "total": total}


async def list_deleted_user_files(
    *,
    user_id: str,
    file_type: Optional[str],
    limit: int,
    offset: int,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    capped_limit = min(limit, 500)
    rows = await entity_file_dao.get_deleted_user_files(
        user_id,
        file_type,
        capped_limit,
        offset,
    )
    total = await entity_file_dao.count_deleted_user_files(user_id, file_type)
    return {"success": True, "items": _normalize_file_rows(rows), "total": total}


async def get_deleted_user_file(
    *,
    file_id: str,
    user_id: str,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    row = await entity_file_dao.get_deleted_user_file(file_id, user_id)
    if not row:
        raise EntityFileNotFound("Deleted file not found or access denied")
    return _normalize_file_rows([row])[0]


async def list_entity_files(
    *,
    entity_type: str,
    entity_id: str,
    file_role: Optional[str],
    limit: int,
    offset: int,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    capped_limit = min(limit, 200)
    result = await entity_file_dao.get_entity_files(
        entity_type,
        entity_id,
        file_role,
        capped_limit,
        offset,
    )
    return {"success": True, **result}


async def link_entity_file(
    *,
    file_id: str,
    entity_type: str,
    entity_id: str,
    file_role: str,
    is_selected: bool,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    row = await entity_file_dao.link_file(
        file_id,
        entity_type,
        entity_id,
        file_role,
        is_selected,
    )
    if not row:
        raise EntityFileNotFound("File not found or deleted")
    return {"success": True, "file": row}


async def select_entity_file(
    *,
    file_id: str,
    entity_type: str,
    entity_id: str,
    file_role: str,
    entity_file_dao: Any,
    logger: Optional[logging.Logger] = None,
    selected_by: Optional[str] = None,
    content_workflow_dao: Any = None,
) -> Dict[str, Any]:
    row = await entity_file_dao.select_file(file_id, entity_type, entity_id, file_role)
    if not row:
        raise EntityFileNotFound("File not found or not linked to entity")

    try:
        await entity_file_dao.sync_legacy_url(entity_type, entity_id, file_role, row["file_url"])
    except Exception as exc:
        if logger:
            logger.warning("同步旧URL字段失败: %s", exc)

    # Keep old file selection and the new type-agnostic selection in one user
    # action.  During a rolling migration this remains best-effort.
    try:
        if content_workflow_dao is None:
            from dao.creative.content_workflow import ContentWorkflowDAO

            content_workflow_dao = ContentWorkflowDAO
        from services.content_workflow_service import (
            list_content_takes,
            normalize_entity_type,
            select_content_take,
            slot_for_file,
        )

        slot = slot_for_file(str(row.get("file_type") or ""), file_role)
        if slot:
            normalized_type = normalize_entity_type(entity_type)
            takes_result = await list_content_takes(
                entity_type=normalized_type,
                entity_id=entity_id,
                slot=slot,
                workflow_dao=content_workflow_dao,
            )
            take = next(
                (item for item in takes_result["items"] if item.get("file_id") == file_id),
                None,
            )
            if take:
                await select_content_take(
                    entity_type=normalized_type,
                    entity_id=entity_id,
                    slot=slot,
                    take_id=take["take_id"],
                    selected_by=selected_by,
                    workflow_dao=content_workflow_dao,
                )
    except Exception as exc:
        if logger:
            logger.warning("统一候选选择同步失败（旧选择已生效）: %s", exc)
    return {"success": True, "file": row}


async def upload_entity_file(
    *,
    content: bytes,
    filename: Optional[str],
    content_type: Optional[str],
    entity_type: Optional[str],
    entity_id: Optional[str],
    file_role: Optional[str],
    episode_id: Optional[str],
    user_id: str,
    save_generated_file_to_db: Callable[..., Any],
    project_id: Optional[str] = None,
    media_library_create_from_file: Optional[Callable[..., Any]] = None,
    logger: Optional[logging.Logger] = None,
) -> Dict[str, Any]:
    original_ext = Path(filename).suffix if filename else ".bin"
    file_type = _detect_file_type(content_type)

    saved = await save_generated_file_to_db(
        content=content,
        file_type=file_type,
        user_id=user_id,
        source="upload",
        entity_type=entity_type,
        entity_id=entity_id,
        file_role=file_role,
        original_ext=original_ext,
        project_id=project_id,
        episode_id=episode_id,
    )

    if file_type in ("image", "video", "audio"):
        try:
            create_from_file = media_library_create_from_file
            if create_from_file is None:
                import media_library_service

                create_from_file = media_library_service.create_from_file
            await create_from_file(
                file_record=saved,
                source="upload",
                episode_id=episode_id,
                source_entity_type=entity_type,
                source_entity_id=entity_id,
                title=(filename or "")[:80] or None,
            )
        except Exception as exc:
            if logger:
                logger.warning("media_library 同步失败 (entity-files upload): %s", exc)

    return {"success": True, "file_id": saved["file_id"], "file_url": saved["file_url"]}


async def soft_delete_entity_file(
    *,
    file_id: str,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    ok = await entity_file_dao.soft_delete(file_id)
    if not ok:
        raise EntityFileNotFound("File not found or deleted")
    return {"success": True}


async def restore_entity_file(
    *,
    file_id: str,
    user_id: str,
    entity_file_dao: Any,
) -> Dict[str, Any]:
    row = await entity_file_dao.restore(file_id, user_id)
    if not row:
        raise EntityFileNotFound("Deleted file not found or access denied")
    return {"success": True, "file": row}


async def hard_delete_entity_file(
    *,
    file_id: str,
    user_id: str,
    risk_ack: bool,
    entity_file_dao: Any,
    node_output_deleter: Optional[Callable[..., Any]] = None,
) -> Dict[str, Any]:
    if not risk_ack:
        raise EntityFileDeleteRiskNotAcknowledged("Permanent deletion must be acknowledged")

    row = await entity_file_dao.get_deleted_user_file(file_id, user_id)
    if not row:
        raise EntityFileNotFound("Deleted file not found or access denied")
    normalized = _normalize_file_rows([row])[0]
    metadata = normalized.get("metadata") if isinstance(normalized.get("metadata"), dict) else {}
    node_output_id = str(metadata.get("node_output_id") or "").strip()
    node_agent_id = str(metadata.get("node_agent_id") or "").strip()
    is_node_local = str(normalized.get("file_path") or "").startswith("agent://") or (
        str(metadata.get("source") or "") == "node_local_output"
    )

    try:
        if is_node_local:
            if not node_output_id or not node_agent_id or node_output_deleter is None:
                raise EntityFilePhysicalDeleteFailed("Local node deletion information is unavailable")
            remote = await node_output_deleter(
                agent_id=node_agent_id,
                output_id=node_output_id,
            )
            deleted = await entity_file_dao.delete_deleted_user_file_record(file_id, user_id)
            if not deleted:
                raise EntityFilePhysicalDeleteFailed("File record changed during permanent deletion")
            freed_bytes = max(0, int((remote or {}).get("freed_bytes") or 0))
        else:
            result = await entity_file_dao.hard_delete(file_id, user_id)
            if not result:
                raise EntityFileNotFound("Deleted file not found or access denied")
            freed_bytes = max(0, int(result.get("freed_bytes") or 0))
    except EntityFileServiceError:
        raise
    except (OSError, RuntimeError, ValueError) as exc:
        raise EntityFilePhysicalDeleteFailed(str(exc)) from exc

    return {"success": True, "file_id": file_id, "freed_bytes": freed_bytes}


async def hard_delete_entity_files_batch(
    *,
    file_ids: list[str],
    user_id: str,
    risk_ack: bool,
    entity_file_dao: Any,
    node_output_deleter: Optional[Callable[..., Any]] = None,
) -> Dict[str, Any]:
    if not risk_ack:
        raise EntityFileDeleteRiskNotAcknowledged("Permanent deletion must be acknowledged")
    unique_ids = [file_id for file_id in dict.fromkeys(file_ids or []) if file_id]
    if len(unique_ids) > 200:
        raise EntityFileBatchTooLarge("Batch hard delete accepts at most 200 files")

    deleted = 0
    freed_bytes = 0
    errors = []
    for file_id in unique_ids:
        try:
            result = await hard_delete_entity_file(
                file_id=file_id,
                user_id=user_id,
                risk_ack=True,
                entity_file_dao=entity_file_dao,
                node_output_deleter=node_output_deleter,
            )
            deleted += 1
            freed_bytes += int(result.get("freed_bytes") or 0)
        except EntityFileServiceError as exc:
            errors.append({"file_id": file_id, "error": str(exc)})
    return {
        "success": not errors,
        "requested": len(unique_ids),
        "deleted": deleted,
        "freed_bytes": freed_bytes,
        "errors": errors,
    }


async def run_entity_file_migration(
    *,
    migration_runner: Optional[Callable[[], Any]] = None,
) -> Dict[str, Any]:
    try:
        if migration_runner is not None:
            recovered = await migration_runner()
            return {"success": True, "recovered": recovered}

        from migrate_existing_files import (
            migrate_assets,
            migrate_storyboard_items,
            migrate_video_segments,
            recover_orphan_files,
        )

        await migrate_storyboard_items()
        await migrate_assets()
        await migrate_video_segments()
        recovered = await recover_orphan_files()
        return {"success": True, "recovered": recovered}
    except Exception as exc:
        raise EntityFileMigrationFailed(str(exc)) from exc
